server/play-server/src/test/scala/sttp/tapir/server/play/PlayServerTest.scala

```
@@ -98,7 +109,61 @@ class PlayServerTest extends TestSuite {
+              .serverLogicSuccess[Future] { case (length, stream) =>
+                Future.successful(length, stream)
+              }
+
+          val components: DefaultPekkoHttpServerComponents = new DefaultPekkoHttpServerComponents {
+            val initialServerConfig: ServerConfig = ServerConfig(port = Some(0), address = "127.0.0.1", mode = Mode.Test)
+
+            val customConf: Configuration =
+              Configuration(
+                ConfigFactory.parseString(s"play { server.pekko.requestTimeout=${requestTimeout.toString} }")
+              )
+            override lazy val serverConfig: ServerConfig =
+              initialServerConfig.copy(configuration = customConf.withFallback(initialServerConfig.configuration))
+            override lazy val actorSystem: ActorSystem = ActorSystem("tapir", defaultExecutionContext = Some(_actorSystem.dispatcher))
+            override lazy val router: Router = Router.from(PlayServerInterpreter().toRoutes(e)).withPrefix("/chunks")
+          }
+
+          def createStream(chunkSize: Int, beforeSendingSecondChunk: FiniteDuration): Stream[IO, Byte] = {
+            val chunk = Chunk.array(Array.fill(chunkSize)('A'.toByte))
+            val initialChunks = Stream.chunk(chunk)
+            val delayedChunk = Stream.sleep[IO](beforeSendingSecondChunk) >> Stream.chunk(chunk)
+            initialChunks ++ delayedChunk
+          }
+
+          val inputStream = createStream(chunkSize, beforeSendingSecondChunk)
+
+          val bind = IO.blocking(components.server)
+          Resource.make(bind)(s => IO.blocking(s.stop()))
+            .map(_.mainAddress.getPort)
+            .use { port =>
+              basicRequest
+                .post(uri"http://localhost:$port/chunks")
+                .contentLength(2 * chunkSize)
+                .streamBody(Fs2Streams[IO])(inputStream)
+                .send(backend)
+                .map{ response =>
+                  response.code shouldBe StatusCode.Ok
```

hm well if this test passes, something is wrong - we set the timeout to 1s, so we should never receive a response if it takes 2s to send it? unless the request timeout is for something else? 

anyway, this doesn't test the scenario from the test case - where the transmission is interrupted half-way because of connection problems; I don't know if we can simulate this in a test case, but using a timeout is a good approximation. But probably a good way to check if we can at all reproduce the bug is to run: a long-running client sender process; a server process; then `kill -9` the client process when it's half-way sending the data, and seeing on the server if received the incomplete data in the server logic

---

I have given it another try as you suggested. There are [playServer](https://github.com/softwaremill/tapir/pull/4214/commits/9402d02a2f735e0c553132f6c6d62687dd0dd183#diff-680e4d576aafc5630a9032b9619d0f1b5d93d8d6f55871da044e8eca33d52897) and [longLastingClient](https://github.com/softwaremill/tapir/pull/4214/commits/9402d02a2f735e0c553132f6c6d62687dd0dd183#diff-91749d496b59d0d54eacb34e5cecb106ecb4495584408b87ce58f4baad8340c4) but I don't know what's wrong with that approach. Suggestions are welcome 💡

---

When you write "I don't know what's wrong with this approach ", do you mean that it works as expected (that is: you run both, interrupt the client brutally after some time, and the server properly closes the connection), or is there something else that's wrong?

---

I'm trying with the following steps:

1. `$ scala-cli run playServer.scala`

```
[playServer-pekko.actor.default-dispatcher-4] INFO org.apache.pekko.event.slf4j.Slf4jLogger - Slf4jLogger started
[application-pekko.actor.default-dispatcher-6] INFO org.apache.pekko.event.slf4j.Slf4jLogger - Slf4jLogger started
[main] INFO play.api.Play - Application started (Dev) (no global state)
[main] INFO play.core.server.NettyServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
```

2. `$ scala-cli run longLastingClient.scala`

```
ByteString(65, 65, 65, 65, 65, 65, 65, 65, 65, 65) 17:46:14.514330
ByteString(65, 65, 65, 65, 65, 65, 65, 65, 65, 65) 17:46:15.534314
ByteString(65, 65, 65, 65, 65, 65, 65, 65, 65, 65) 17:46:16.553033
ByteString(65, 65, 65, 65, 65, 65, 65, 65, 65, 65) 17:46:17.573135
....
```

3. server side

```
Received 10000 bytes, Source(SourceShape(Map.out(1467551936))) bytes in total
```

4.  `$ ps aux | grep longLastingClient | awk '{print $2}' | head -n 1 | xargs kill -9`

5.  and nothing new (error/exception) on server side 🤔

---

I think the lack of exception/error might be the problem here ;) but first there are two problems in the code:
1. in the client code, you claim to send 10000 bytes in the content-length, but in fact you might send more?
2. in the server code, you do `stream.map(_.length)`, which just creates a `Source[Long]`, that is a description of a stream that produces lenghts of received byte-strings (byte chunks). You never run (receive) the stream, and that's where you'd expect to see errors (when the stream is being run)

---

🤦‍♂️ added pointed adjustments and the result is the following :

1. 
```
 $ scala-cli run playServer.scala
[playServer-pekko.actor.default-dispatcher-4] INFO org.apache.pekko.event.slf4j.Slf4jLogger - Slf4jLogger started
[application-pekko.actor.default-dispatcher-4] INFO org.apache.pekko.event.slf4j.Slf4jLogger - Slf4jLogger started
[main] INFO play.api.Play - Application started (Dev) (no global state)
[main] INFO play.core.server.NettyServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
```

2. 
```
$ scala-cli run longLastingClient.scala
Chunk 1 sent 13:00:13.650915
Chunk 2 sent 13:00:14.668797
Chunk 3 sent 13:00:15.690020
Chunk 4 sent 13:00:16.710184
Chunk 5 sent 13:00:17.730033
Chunk 6 sent 13:00:18.748416
Chunk 7 sent 13:00:19.769451
Chunk 8 sent 13:00:20.789355
```

3. killing client

4. server side

```
Stream finished: 800/10000 transmitted
```

😞

---

So, I suppose, that's the problem as originally reported?

---

We can now try to fix it, and/or reproduce using a test :)

---

So, I suppose that we should return an error with a descriptive message i.e. `java.lang.IllegalStateException: Expected `Content-Length: 10000` bytes, but only 800 were written`, right?

---

In my case I have an endpoint accepting JSON
`endpoint
      .in("permissions")
      .in(jsonBody[PermissionsInput])
      .out(statusCode)`

I used `netcat` to send a partial request

PUT /permissions HTTP/1.1
Host: example.com
Content-Type: application/json
Content-Length: 500

{
  "permissions": [
    {
      "id": "sleep",
      "platform": "health-kit",

I expect payload to be deserialized only after receiving the whole body

What I see instead

`
2025-04-17 17:35:24,749 WARN [scala-execution-context-global-44] Server$ - [Logger.scala:46] {} Request: PUT /permissions, handled by: PUT /permissions, took: 5498ms; decode failure: Error({
  "permissions": [
    {
      "id": "sleep",
      "platform": "health-kit",

,JsonDecodeException(List(JsonError(exhausted input,List())),io.circe.Errors)), on input: {body as application/json (UTF-8)}; response: 400 
`

---

Netty

---

@akozich could you maybe provide a full scenario which would allow us to reproduce the problem? including the shell commands

---

Here's the server
```
import io.circe.derivation.Configuration
import io.circe.derivation.ConfiguredCodec
import sttp.model.StatusCode
import sttp.tapir.*
import sttp.tapir.Schema
import sttp.tapir.json.circe.*
import sttp.tapir.server.netty.NettyFutureServer
import sttp.tapir.server.netty.NettyFutureServerOptions

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

given Configuration = Configuration.default.withSnakeCaseMemberNames.withDefaults

case class PermissionsInput(
  permissions: Seq[PermissionInput]
) derives ConfiguredCodec, Schema

case class PermissionInput(
  platform: String,
  id: String,
  accessType: String,
  state: String
) derives ConfiguredCodec, Schema

object Server {

  private val permissionsEndpoint =
    endpoint.put
      .in("permissions")
      .in(jsonBody[PermissionsInput])
      .out(statusCode)

  private val permissionsServerEndpoint = permissionsEndpoint
    .serverLogicSuccess { (permissionsInput: PermissionsInput) =>
      println(permissionsInput.permissions.size)
      Future.successful(StatusCode.Created)
    }

  private val server = NettyFutureServer()
    .host("0.0.0.0")
    .options(
      NettyFutureServerOptions.customiseInterceptors.serverLog(
        NettyFutureServerOptions.defaultServerLog.logAllDecodeFailures(true).logWhenHandled(true)
      ).options
    )
    .port(8080)
    .addEndpoints(permissionsServerEndpoint :: Nil)

  def main(args: Array[String]): Unit = {
    server.start().andThen {
      case scala.util.Success(_) =>
        println("Server started on port 8080")
      case scala.util.Failure(exception) =>
        println(s"Failed to start server: ${exception.getMessage}")
    }
  }

}
```

And here's the client sending partial requets

```
import java.io.PrintWriter
import java.net.Socket

object TruncatedRequestsSender extends App {

  val socket = new Socket("localhost", 8080)
  val out = new PrintWriter(socket.getOutputStream, true)

  out.print(
    """PUT /permissions HTTP/1.1
      |Host: localhost:8080
      |Content-Type: application/json
      |Content-Length: 500
      |
      |{
      |  "permissions": [
      |    {
      |      "id": "sleep",""".stripMargin
  )

  out.close()
  socket.close()
}
```

You'll see in the logs

```
13:00:57.931 [DEBUG] s.t.s.n.NettyFutureServerOptions$ - Request: PUT /permissions, handled by: PUT /permissions, took: 2ms; decode failure: Error({
  "permissions": [
    {
      "id": "sleep",,JsonDecodeException(List(JsonError(exhausted input,List())),io.circe.Errors)), on input: {body as application/json (UTF-8)}; response: 400
io.circe.Errors: null
```

which means decoder was trying to decode the partial body

---

In this scenario I do get exception from time to time.

```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
	Hint: 'HttpStreamsServerHandler#0-body-publisher' will handle the message from this point.
	io.netty.handler.codec.http.DefaultHttpContent.touch(DefaultHttpContent.java:86)
	io.netty.handler.codec.http.DefaultHttpContent.touch(DefaultHttpContent.java:25)
        ...
```

If the request is not full the reference to the byteBuf is stored inside `SimpleSubscriber`. Then, `Future` is returned from it. If the reference to the  Future is garbage collected before receiving the last element from the stream the reference to the byte buffer is just removed without releasing the byte buffer

---

Thanks for the reproducing test case :). Though if you are closing the connection & socket on the client side, the server side receives a "data complete signal" (even though it's less than the declared amount), and continues to process it. So `SimpleSubscriber.onComplete` is called, which gathers all buffers into a byte array and passes it for parsing. Hence I would say that the current behavior is at least somewhat correct. What would you expect instead?

As for the `LEAK` warning, that must be a bug somewhere. Although this unfortunately I can't reproduce. I tried following your line of thought, but I can't see where the future is GCed before being fully handled?

---

I would expect received bytes to be skipped (not pushed down) and `400 Bad request` to be returned immediately. 

Not sure if I can reproduce LEAK with this sample code but in my production code I can easily reproduce it by:

1. Repeating partial requests in the loop
2. Limiting the heap size for the server

After doing ~200-300 requests I get 

```
2025-04-22 09:32:38,452 ERROR[KQueueEventLoopGroup-2-12] i.n.u.ResourceLeakDetector - [ResourceLeakDetector.java:337]  LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
	Hint: 'HttpStreamsServerHandler#0-body-publisher' will handle the message from this point.
	io.netty.handler.codec.http.DefaultHttpContent.touch(DefaultHttpContent.java:86)
	io.netty.handler.codec.http.DefaultHttpContent.touch(DefaultHttpContent.java:25)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:115)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:346)
	org.playframework.netty.http.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:194)
	org.playframework.netty.http.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:165)
	org.playframework.netty.http.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:96)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:107)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:359)
	io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1429)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
	io.netty.channel.kqueue.AbstractKQueueStreamChannel$KQueueStreamUnsafe.readReady(AbstractKQueueStreamChannel.java:543)
	io.netty.channel.kqueue.AbstractKQueueChannel$AbstractKQueueUnsafe.readEOF(AbstractKQueueChannel.java:552)
	io.netty.channel.kqueue.AbstractKQueueChannel$AbstractKQueueUnsafe.handle(AbstractKQueueChannel.java:435)
	io.netty.channel.kqueue.KQueueIoHandler$DefaultKqueueIoRegistration.handle(KQueueIoHandler.java:396)
	io.netty.channel.kqueue.KQueueIoHandler.processReady(KQueueIoHandler.java:181)
	io.netty.channel.kqueue.KQueueIoHandler.run(KQueueIoHandler.java:237)
	io.netty.channel.SingleThreadIoEventLoop.runIo(SingleThreadIoEventLoop.java:204)
	io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:175)
	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1073)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:1583)
```

---

Thanks; could you maybe create two new issues: (1) return 400 if not all declared bytes are received, and (2) leak when running many incomplete requests in Netty?

I think these are separate from the one discussed in this PR

For (1) I'd also like to verify other servers, what's the "living standard"

---

I know akka/pekko is doing exactly this. When I call `toStrictEntity`

`HttpMessageParser` returns `EntityStreamError(ErrorInfo(
        "Entity stream truncation. The HTTP parser was receiving an entity when the underlying connection was closed unexpectedly."))`. 

`toStrictEntity` throws `throw IllegalRequestException(StatusCodes.BadRequest, info)` which is then handled inside `ExceptionHandler.default`

---

1. I updated the title of the previously created bug https://github.com/softwaremill/tapir/issues/4169 for the first problem
2. Created a new bug for the LEAK https://github.com/softwaremill/tapir/issues/4539